src/kits/serial/native/linux/serial_SerialPort_linux.c

```diff
@@ -0,0 +1,453 @@
+//
+// Copyright (c) 2016 Andrey Skvortsov <andrej.skvortzov@gmail.com>
+// Licensed under the Academic Free License version 3.0
+//
+
+
+#include "sedona.h"
+#include "serial_SerialPort.h"
+
+#include <termios.h>
+#include <sys/ioctl.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <linux/serial.h>
+#include <fcntl.h>
+#include <unistd.h>
+
+bool debug = false;
+
+struct SerialPortLinux {
+	int fd;
+
+};
+
+#define maxPortNum 32
+static struct SerialPortLinux portHandles[maxPortNum+1];
+
+int __attribute__((weak)) getMaxPortNum()
+{
+	return maxPortNum;
+}
+
+const char* __attribute__((weak)) getPortName(int portNum)
+{
+	static char port[32];
+	snprintf(port, sizeof(port), "/dev/ttyS%d",portNum);
+	return port;
+}
+
+
+static inline bool checkPortNum(int portNum)
+{
+	if (portNum > maxPortNum) {
+		fprintf(stderr, "portNum (%d) is greater than last supported port (%d)\n",
+			portNum,
+			maxPortNum);
+		return false;
+	}
+	return ((portNum>0) && (portNum<=getMaxPortNum()));
```

Comment:
maxPortNum defines maximal number of file handler.
getMaxPortNum is weak function and it could be reimplemented by vendor and overwritten at link time. It could be used to implement any vendor specific logic. For example, dynamic port handling. First maxPortNum check is just extra check, just in case vendor's getMaxPortNum is buggy for some reason.

Reply:
ok, I see. just learn it from [wikipedia](https://en.wikipedia.org/wiki/Weak_symbol), good feature